File: backend/app/services/drawing_service.py
import json
import uuid
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from fastapi import HTTPException

from app.schemas.drawing import Drawing, DrawingScene
from app.services.note_service import note_exists
from app.services.workspace_service import get_workspace_dir

METADATA_FILENAME = "drawings.jsonl"
EMPTY_SCENE: dict[str, Any] = {"elements": [], "appState": {}, "files": {}}


def _metadata_path(workspace_dir: Path) -> Path:
    return workspace_dir / METADATA_FILENAME


def _scene_path(workspace_dir: Path, drawing_id: str) -> Path:
    return workspace_dir / "drawings" / f"{drawing_id}.excalidraw"


def _read_all(workspace_dir: Path) -> list[dict]:
    path = _metadata_path(workspace_dir)
    if not path.exists():
        return []
    return [json.loads(line) for line in path.read_text().splitlines() if line.strip()]


def _write_all(workspace_dir: Path, entries: list[dict]) -> None:
    _metadata_path(workspace_dir).write_text("".join(json.dumps(entry) + "\n" for entry in entries))


def _find(entries: list[dict], drawing_id: str) -> dict:
    # Same reasoning as note_service._find: drawing_id is only ever checked
    # against this per-workspace index, which only ever contains ids we
    # generated ourselves, so it can't be used to escape the drawings dir.
    for entry in entries:
        if entry["id"] == drawing_id:
            return entry
    raise HTTPException(status_code=404, detail="Drawing not found")


def _to_drawing(entry: dict) -> Drawing:
    return Drawing(**entry)


def _require_note(workspace_id: str, note_id: str) -> None:
    if not note_exists(workspace_id, note_id):
        raise HTTPException(status_code=404, detail="Note not found")


def list_drawings(workspace_id: str, note_id: str) -> list[Drawing]:
    _require_note(workspace_id, note_id)
    workspace_dir = get_workspace_dir(workspace_id)
    entries = [e for e in _read_all(workspace_dir) if e["note_id"] == note_id]
    entries.sort(key=lambda e: e["updated_at"], reverse=True)
    return [_to_drawing(e) for e in entries]


def get_drawing(workspace_id: str, drawing_id: str) -> DrawingScene:
    workspace_dir = get_workspace_dir(workspace_id)
    entry = _find(_read_all(workspace_dir), drawing_id)
    scene = json.loads(_scene_path(workspace_dir, drawing_id).read_text())
    return DrawingScene(**entry, scene=scene)


def create_drawing(workspace_id: str, note_id: str, title: str) -> DrawingScene:
    _require_note(workspace_id, note_id)
    title = title.strip()
    if not title:
        raise HTTPException(status_code=422, detail="Drawing title cannot be empty")

    workspace_dir = get_workspace_dir(workspace_id)
    drawing_id = uuid.uuid4().hex
    now = datetime.now(UTC).isoformat()
    entry = {
        "id": drawing_id,
        "note_id": note_id,
        "title": title,
        "created_at": now,
        "updated_at": now,
    }

    entries = _read_all(workspace_dir)
    entries.append(entry)
    _write_all(workspace_dir, entries)
    _scene_path(workspace_dir, drawing_id).write_text(json.dumps(EMPTY_SCENE))

    return DrawingScene(**entry, scene=EMPTY_SCENE)


def save_scene(workspace_id: str, drawing_id: str, scene: dict[str, Any]) -> Drawing:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, drawing_id)
    entry["updated_at"] = datetime.now(UTC).isoformat()
    _write_all(workspace_dir, entries)
    _scene_path(workspace_dir, drawing_id).write_text(json.dumps(scene))
    return _to_drawing(entry)


def rename_drawing(workspace_id: str, drawing_id: str, title: str) -> Drawing:
    title = title.strip()
    if not title:
        raise HTTPException(status_code=422, detail="Drawing title cannot be empty")

    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, drawing_id)
    entry["title"] = title
    entry["updated_at"] = datetime.now(UTC).isoformat()
    _write_all(workspace_dir, entries)
    return _to_drawing(entry)


def delete_drawing(workspace_id: str, drawing_id: str) -> None:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, drawing_id)
    entries.remove(entry)
    _write_all(workspace_dir, entries)
    _scene_path(workspace_dir, drawing_id).unlink(missing_ok=True)

File: apps/api/src/services/drawingService.ts
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { readJsonl, writeJsonl } from '../utils/fileDb.js';
import { getWorkspaceDir } from './workspaceService.js';
import { noteExists, readAllNotes } from './noteService.js';
import { Drawing, DrawingScene } from '@ticks/types';

const METADATA_FILENAME = 'drawings.jsonl';
const EMPTY_SCENE = { elements: [], appState: {}, files: {} };

function metadataPath(workspaceDir: string): string {
  return path.join(workspaceDir, METADATA_FILENAME);
}

function scenePath(workspaceDir: string, drawingId: string): string {
  return path.join(workspaceDir, 'drawings', `${drawingId}.excalidraw`);
}

function readAllDrawings(workspaceDir: string): any[] {
  return readJsonl(metadataPath(workspaceDir));
}

function writeAllDrawings(workspaceDir: string, entries: any[]): void {
  writeJsonl(metadataPath(workspaceDir), entries);
}

function findDrawing(entries: any[], drawingId: string): any {
  const entry = entries.find(e => e.id === drawingId);
  if (!entry) {
    throw { status: 404, message: 'Drawing not found' };
  }
  return entry;
}

function requireNote(workspaceId: string, noteId: string): void {
  if (!noteExists(workspaceId, noteId)) {
    throw { status: 404, message: 'Note not found' };
  }
}

export function listDrawings(workspaceId: string, noteId: string | null | undefined, includeAll = false): Drawing[] {
  if (noteId !== null && noteId !== undefined) {
    requireNote(workspaceId, noteId);
  }
  const workspaceDir = getWorkspaceDir(workspaceId);
  const rawEntries = readAllDrawings(workspaceDir);
  
  let entries = includeAll
    ? rawEntries
    : rawEntries.filter(e => e.note_id === noteId);

  // Map note titles
  let noteTitles: Record<string, string> = {};
  try {
    const notes = readAllNotes(workspaceDir);
    for (const n of notes) {
      noteTitles[n.id] = n.title;
    }
  } catch (e) {
    // Ignore
  }

  for (const e of entries) {
    if (e.note_id) {
      e.note_title = noteTitles[e.note_id] || null;
    }
  }

  return entries
    .sort((a, b) => new Date(b.updated_at).getTime() - new Date(a.updated_at).getTime());
}

export function getDrawing(workspaceId: string, drawingId: string): DrawingScene {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entry = findDrawing(readAllDrawings(workspaceDir), drawingId);
  const filePath = scenePath(workspaceDir, drawingId);
  const scene = fs.existsSync(filePath) ? JSON.parse(fs.readFileSync(filePath, 'utf8')) : EMPTY_SCENE;
  return { ...entry, scene };
}

export function createDrawing(workspaceId: string, noteId: string | null | undefined, title: string): DrawingScene {
  if (noteId !== null && noteId !== undefined) {
    requireNote(workspaceId, noteId);
  }
  const trimmedTitle = title.trim();
  if (!trimmedTitle) {
    throw { status: 422, message: 'Drawing title cannot be empty' };
  }

  const workspaceDir = getWorkspaceDir(workspaceId);
  const drawingId = crypto.randomUUID().replace(/-/g, '');
  const now = new Date().toISOString();

  const entry = {
    id: drawingId,
    note_id: noteId || null,
    title: trimmedTitle,
    created_at: now,
    updated_at: now
  };

  const entries = readAllDrawings(workspaceDir);
  entries.push(entry);
  writeAllDrawings(workspaceDir, entries);

  fs.writeFileSync(scenePath(workspaceDir, drawingId), JSON.stringify(EMPTY_SCENE), 'utf8');

  return { ...entry, scene: EMPTY_SCENE };
}

export function saveScene(workspaceId: string, drawingId: string, scene: Record<string, any>): Drawing {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllDrawings(workspaceDir);
  const entry = findDrawing(entries, drawingId);
  entry.updated_at = new Date().toISOString();
  writeAllDrawings(workspaceDir, entries);
  fs.writeFileSync(scenePath(workspaceDir, drawingId), JSON.stringify(scene), 'utf8');
  return entry;
}

export function renameDrawing(workspaceId: string, drawingId: string, title: string): Drawing {
  const trimmedTitle = title.trim();
  if (!trimmedTitle) {
    throw { status: 422, message: 'Drawing title cannot be empty' };
  }

  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllDrawings(workspaceDir);
  const entry = findDrawing(entries, drawingId);
  entry.title = trimmedTitle;
  entry.updated_at = new Date().toISOString();
  writeAllDrawings(workspaceDir, entries);
  return entry;
}

export function deleteDrawing(workspaceId: string, drawingId: string): void {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllDrawings(workspaceDir);
  const entry = entries.find(e => e.id === drawingId);
  if (entry) {
    const index = entries.indexOf(entry);
    if (index > -1) {
      entries.splice(index, 1);
    }
    writeAllDrawings(workspaceDir, entries);
  }
  const filePath = scenePath(workspaceDir, drawingId);
  if (fs.existsSync(filePath)) {
    fs.unlinkSync(filePath);
  }
}
